[SPARK-30511][SPARK-28403][CORE] Don't treat failed/killed speculative tasks as pending in ExecutorAllocationManager

linzebing opened the pull request:
### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Currently, when speculative tasks fail/get killed, they are still considered as pending and count towards the calculation of number of needed executors. To be more accurate: `stageAttemptToNumSpeculativeTasks(stageAttempt)` is incremented on onSpeculativeTaskSubmitted, but never decremented.  `stageAttemptToNumSpeculativeTasks -= stageAttempt` is performed on stage completion. **This means Spark is marking ended speculative tasks as pending, which leads to Spark to hold more executors that it actually needs!**

This PR fixes this issue by updating `stageAttemptToSpeculativeTaskIndices` and  `stageAttemptToNumSpeculativeTasks` on speculative tasks completion.  This PR also addresses some other minor issues: scheduler behavior after receiving an intentionally killed task event; try to address [SPARK-28403](https://issues.apache.org/jira/browse/SPARK-28403).


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
This has caused resource wastage in our production with speculation enabled. With aggressive speculation, we found data skewed jobs can hold hundreds of idle executors with less than 10 tasks running.

An easy repro of the issue (`--conf spark.speculation=true --conf spark.executor.cores=4 --conf spark.dynamicAllocation.maxExecutors=1000` in cluster mode):
```
val n = 4000
val someRDD = sc.parallelize(1 to n, n)
someRDD.mapPartitionsWithIndex( (index: Int, it: Iterator[Int]) => {
if (index < 300 && index >= 150) {
    Thread.sleep(index * 1000) // Fake running tasks
} else if (index == 300) {
    Thread.sleep(1000 * 1000) // Fake long running tasks
}
it.toList.map(x => index + ", " + x).iterator
}).collect
```
You will see when running the last task, we would be hold 38 executors (see below), which is exactly (152 + 3) / 4 = 38.
![image](https://user-images.githubusercontent.com/9404831/72469112-9a7fac00-3793-11ea-8f50-74d0ab7325a4.png)

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Added a comprehensive unit test.

Test with the above repro shows that we are holding 2 executors at the end
![image](https://user-images.githubusercontent.com/9404831/72469177-bbe09800-3793-11ea-850f-4a2c67142899.png)

Commit:
Don't mark ended speculative tasks as pending

linzebing commented:
@jiangxb1987 @vanzin @cloud-fan

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

Ngone51 reviewed (commented):
> To be more accurate: `stageAttemptToNumSpeculativeTasks(stageAttempt)` is incremented on onSpeculativeTaskSubmitted, but never decremented.  `stageAttemptToNumSpeculativeTasks -= stageAttempt` is performed on stage completion. This means Spark is marking ended speculative tasks as pending, which leads to Spark to hold more executors that it actually needs!

I think `stageAttemptToNumSpeculativeTasks` is designed not to decrease intentionally. And `ExecutorAllocationManager` use:
https://github.com/apache/spark/blob/3584d849438ad48ff54af3c982c124a8443dc590/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L657-L661

to count pending speculative tasks rather than `stageAttemptToNumSpeculativeTasks` only.

linzebing commented:
@Ngone51 : Let's say we have one speculative task. When a speculative task gets intentionally killed because the original task succeeds, then `pendingSpeculativeTasks` will be 1. However, it should be zero. In my PR, the two data structures are given a very clear meaning:
`stageAttemptToNumSpeculativeTasks`: all the pending and running speculative tasks
`stageAttemptToSpeculativeTaskIndices`: task indexes of all the running speculative tasks.

But thanks for pointing out, I'm editing the PR description to be more accurate.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
+        // ensure we have enough resources to run the resubmitted task, we need to mark the
+        // scheduler as backlogged again if it's not already marked as such (SPARK-8366)
+        taskEnd.reason match {
+          case Success | _: TaskKilled =>
```
Why do we want to include `TaskKilled` here? It's not quite straight forward to me

linzebing replied:
Inside the the brackets, there are two things:
```
if (totalPendingTasks() == 0) {
    allocationManager.onSchedulerBacklogged()
}
```
This one is straightforward. If a task is intentionally killed, then we don't expect this task to be resubmitted again, and we don't need to mark the scheduler as backlogged.
```
if (!taskEnd.taskInfo.speculative) {
    stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
}
```
If a non-speculative task is intentionally killed, it means the speculative task has succeeded, and no further task of this task index will be resubmitted. In this case, the task index is completed and we shouldn't remove it from `stageAttemptToTaskIndices`. Otherwise, we will have a pending non-speculative task for the task index.

tgravescs replied:
I think it would be good to add a comment under the taskkilled explaining this

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
-            stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          }
+
+        if (taskEnd.taskInfo.speculative) {
```
thanks for fixing this, we made a mistake previously.

Ngone51 replied:
I think `stageAttemptToSpeculativeTaskIndices` also includes successful finished speculative tasks? Remove a successful speculative task will count it into pending speculative tasks again.

linzebing replied:
Note we already remove pending task `stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1`. @Ngone51

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -263,9 +263,15 @@ private[spark] class ExecutorAllocationManager(
+    if (listener.pendingSpeculativeTasks > 0 && tasksPerExecutorForFullParallelism > 1) {
+      // If we have pending speculative tasks, allocate one more executor to satisfy the
+      // locality requirements of speculative tasks
+      maxNeeded + 1
```
This could lead to wast of executors ?

linzebing replied:
As specified in the comments, this is to satisfy the locality requirements of speculative tasks. Let's say we have 1 normal task and 1 speculative task, in this case we should allocate 2 executors instead of 1.

tgravescs replied:
I typoed my example should be 1000 tasks with 2 tasks per executor, would results in 500 executors, we don't need to add one more

linzebing replied:
Let's use your 1000 tasks (with 2 tasks/executor) example. The dynamic allocation scheduler currently doesn't track the underlying distribution of tasks on different executors. So it's totally possible that we have 999 tasks running, then on one executor we will have a single task running. And we have a pending speculative task for the same task index as this single task, then the speculative task can't be launched in this case. Doing a "+1" is a very simple way to address this situation.

tgravescs replied:
right but you can't guarantee that the +1 will get your speculative task or that the new executor won't end up on the same host. In our example case of all 999 already running it would give it a better chance if the 1 that finished happened to be on the same executor but what are the odds only 1 finished and it finished exactly on the same host?  Generally the reason you speculate is that you are assuming something on that host is wrong or slow so you want to run it somewhere else.  If that is the case that executor is likely to stay loaded up while other ones finish. If you have 500 executors already and your speculation configs are sane, the odds of another task finishing on a different executor are pretty good. If you don't have all your regular tasks finished, they could easily be put on the +1 host as well before your speculative has a chance to be scheduled on it and you could end up with only the 1 your counter part is on.  Since we don't have them integrated well its all just a guess/heuristic.  

I'm ok with leaving this logic to +1 like you have it (with expanded comment) since that is what it was but at the same time it can be a huge waste or resources. If every application has speculative on and you have 1000 apps running, you could easily be wasting 1000's of containers when you already have enough resources to run within.
We could definitely come up with better heuristics that don't waste so much but can be done separately.
Also can you add in a debug log statement here when we do the +1

linzebing replied:
Thanks! You have made excellent points: the +1 gives it a better chance but not a guarantee, and the slow hosts are more likely to be loaded up. I'm OK with doing a +1 only when maxNeeded = 1, will address it.

tgravescs replied:
another option would be to do the logic for speculative separately instead of lumping it with the normal tasks.  This might be an in between the 2 options. You could for instance calculate how many you need for regular tasks, then calculate separately how many executors for speculative, if  speculative doesn't add another executor on top of the normal tasks then add one.  That would help with us double counting.
I'm fine with any of these as we can follow up to make it smarter in the future.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -263,9 +263,15 @@ private[spark] class ExecutorAllocationManager(
-      .toInt
+    val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
+      tasksPerExecutorForFullParallelism).toInt
+    if (listener.pendingSpeculativeTasks > 0 && tasksPerExecutorForFullParallelism > 1) {
```
Why do we need to verify `tasksPerExecutorForFullParallelism ` here ?

linzebing replied:
Because if we are using 1 task per executor, we don't need to allocate an extra executor for locality requirements.

jiangxb1987 commented:
It seems you are also trying to address https://issues.apache.org/jira/browse/SPARK-28403 ? If that's the case maybe you want to also add `[SPARK-28403]` to the title?
Also, the title should describe the behavior after this PR, not the problem in current code base.

jiangxb1987 commented:
cc @janewangfb

Ngone51 commented:
> Let's say we have one speculative task. When a speculative task gets intentionally killed because the original task succeeds, then pendingSpeculativeTasks will be 1. However, it should be zero. 

Ok, I got your point.

Ngone51 commented:
Seem likes the same problem also exists for normal task when speculative task finished before normal task?

Is it possible to check whether there's another task attempt has succeed when we receive a failed taskEnd event. e.g. ask for TaskSchedulerImpl/TaskSetManager or just record those successful tasks in `ExecutorAllocationManager`.

linzebing commented:
> Seem likes the same problem also exists for normal task when speculative task finished before normal task?
> 
> Is it possible to check whether there's another task attempt has succeed when we receive a failed taskEnd event. e.g. ask for TaskSchedulerImpl/TaskSetManager or just record those successful tasks in `ExecutorAllocationManager`.

If a speculative task finished before the normal task, then the normal task will be killed. I have addressed this case in this PR, see explanation in https://github.com/apache/spark/pull/27223#discussion_r368204585

linzebing commented:
@jiangxb1987  Did I address your concerns? I'm happy to address any further comments.
Maybe @vanzin can take a look here as well.

tgravescs commented:
this looks like the same as https://issues.apache.org/jira/browse/SPARK-27082, linked the jiras

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
+
+        if (taskEnd.taskInfo.speculative) {
+          stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove{taskIndex}}
+          stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1
```
one question I actually have here is what happens on a speculative task failure.  The scheduler will try to rerun but we may not be accounting for it if we have removed, I need to look at the scheduler logic a bit more.

linzebing replied:
When a speculative task fails while the original task is running, it will not be resubmitted (because we have one copy running, see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L289-L293), but a new speculative task will be launched later in a future speculation check cycle. The code works correctly in this case, as we updated both `stageAttemptToSpeculativeTaskIndices` and `stageAttemptToNumSpeculativeTasks`, so this failed speculative task will just behave as it never existed. I can add this to my test case.

@tgravescs, @jiangxb1987: Note that the current design (with or without this PR) will launch more-than-needed executors when **the original task fails when the speculative task is running**. In this case, the original task will not be resubmitted (same as above because we have one copy running). Speculation check will later launch one more speculative task on the already running speculative task. In this case, **we have 2 speculative tasks for the same task index**. In this case, we will still launch more than needed executors.

tgravescs replied:
thanks for the explanation, its been a while since I looked at the speculative handling in the scheduler.   I think the failed task case with speculative is ok. I think the problem is the scheduler isn't  tracking those task as speculative or not after the initial launch.  I would rather error on the side of asking for to many and the failure case should be less frequent.  that is a lot more complex case along with making the dynamic allocation manager smarter about it.  If you can file that other jira I mentioned above to follow up, if you could put those details in there as well that would be great.

jiangxb1987 replied:
sorry I don't follow the logic. According to `TSM.dequeueTask()`:
>     // Tries to schedule a regular task first; if it returns None, then schedules
>     // a speculative task
It would first try to schedule a regular task (which would return None in this case), then it would try to schedule a speculative task. Since we have only the original task running, it shall schedule one speculative task in this case. Am I missing something? I also searched the test case but didn't locate the place describing the behavior.

linzebing replied:
> sorry I don't follow the logic. According to `TSM.dequeueTask()`:
> 
> > ```
> > // Tries to schedule a regular task first; if it returns None, then schedules
> > // a speculative task
> > ```
> 
> It would first try to schedule a regular task (which would return None in this case), then it would try to schedule a speculative task. Since we have only the original task running, it shall schedule one speculative task in this case. Am I missing something? I also searched the test case but didn't locate the place describing the behavior.

@jiangxb1987 Let's make sure we are on the same page. I assume you are talking about the case on `speculation task failure`. As I have explained above, when a speculative task fails, `handleFailedTask` will be invoked, and we will have `addPendingTask(index)` at L894. This will add a pending task for the failed task index (with `speculatable` default to false).

Then from `TSM.dequeueTask()`, we have a call chain of `dequeueTaskHelper` -> `dequeueTaskFromList`, see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L289-L293: as we already have one copy running (the original task), the enqueued pending task will not be resubmitted here. Instead, in a future speculation cycle, we will see the original task running too slow and will try to speculate again on the original task from there.

linzebing replied:
I guess you might be questioning on the code itself. As I have explained above, a failed speculative task will not be resubmitted, but instead a new speculative task will be launched in a future speculation cycle.

jiangxb1987 replied:
Yes, there seems to be a bug at L894, a failed speculative task should be add to the pendingSpeculatableTasks. But as you explained, your fix is correct here.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
+
+        // If the task failed (not intentionally killed), we expect it to be resubmitted later. To
+        // ensure we have enough resources to run the resubmitted task, we need to mark the
+        // scheduler as backlogged again if it's not already marked as such (SPARK-8366)
```
I think we should move this comment down by the call to onSchedulerBacklogged

tgravescs commented on core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala:
```diff
@@ -263,6 +263,99 @@ class ExecutorAllocationManagerSuite extends SparkFunSuite {
+      createTaskInfo(taskId = i, taskIndex = i - 10, executorId = s"${i / 4}", speculative = true)}
+      .foreach { info => post(SparkListenerTaskStart(0, 0, info))}
+    adjustRequestedExecutors(manager)
+    assert(numExecutorsTarget(manager) == 5) // At this point, we still have 6 executors running
```
wait why did this number change to 5? none of the tasks finished between this and the previous check

linzebing replied:
@tgravescs The number of pending speculative tasks becomes 0 now. So now we are returning `maxNeeded` instead of `maxNeeded+1`, which is 5.

tgravescs replied:
ok makes sense, can you add comment.

Commit:
address comments

linzebing commented:
@tgravescs I have addressed your commented, and improved the test case to include cases where speculative/original tasks can fail.

Regarding the JIRA you'd like me to create, are you referring to improve the logic of speculation task tracking in `TaskSetManager.scala`? I don't fully understand the details you want me to fill in. Currently I can only think of it's possible for us to launch two speculative tasks for the same task index, which adds complexity to speculative task tracking in `ExecutorAllocationManager`.

tgravescs reviewed (approved):
thanks for the updates, changes look good.  I can file the jira as a followup to improve the handling if we need to.  it would require more tracking or knowledge of things.

linzebing commented:
@tgravescs Thanks! Since this is my first contribution, I need one admin to verify this patch and a committer to authorize testing with a comment "ok to test”. I plan to contribute again in the future, maybe the community can add me to whitelist as well.

@jiangxb1987: do you want to give a second look here? I'm happy to take any further comments.

tgravescs commented:
ok to test

SparkQA commented:
**[Test build #117489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117489/testReport)** for PR 27223 at commit [`3ea1175`](https://github.com/apache/spark/commit/3ea1175a4d7ccf71095a48945fa8946558b83c21).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22247/
Test PASSed.

SparkQA commented:
**[Test build #117489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117489/testReport)** for PR 27223 at commit [`3ea1175`](https://github.com/apache/spark/commit/3ea1175a4d7ccf71095a48945fa8946558b83c21).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117489/
Test PASSed.

jiangxb1987 commented:
LGTM

linzebing commented:
@jiangxb1987 Anything still needed from me?

tgravescs commented:
should be good, I'll commit this tomorrow if no others comments.

tgravescs commented:
merged this, thanks @linzebing

Ngone51 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -614,18 +615,30 @@ private[spark] class ExecutorAllocationManager(
+              // later. To ensure we have enough resources to run the resubmitted task, we need to
+              // mark the scheduler as backlogged again if it's not already marked as such
+              // (SPARK-8366)
+              allocationManager.onSchedulerBacklogged()
```
Previously, we always call `onSchedulerBacklogged` when `totalPendingTasks() == 0` for both speculative and non-speculative tasks. But according to this PR, we should not do this for speculative tasks now. Right? @linzebing

linzebing replied:
As explained in the comment, if a task is intentionally killed, we should not mark the scheduler as backlogged.

If a speculative task fails, while it will not be directly resubmitted, a new speculative task will be launched in next speculation cycle. So it's OK for us to mark the scheduler as backlogged in this case.

Ngone51 replied:
> If a speculative task fails, while it will not be directly resubmitted, a new speculative task will be launched in next speculation cycle. So it's OK for us to mark the scheduler as backlogged in this case.

Hmm...but the new speculative task may not be launched if the normal task finish. And even if it launched, `ExecutorAllocationManager` could still handle it by receiving `SparkListenerSpeculativeTaskSubmitted`. Though, calling `onSchedulerBacklogged` may could reserve executor resource more early to reduce the delay. Fine!

linzebing replied:
The speculation interval is default to 100ms, so a new speculative task will launched almost instantly.

Ngone51 replied:
make sense.
